RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,268 @@
+# Unit checking
+
+Unit checking is the process of inferring the units of Modelica variables, and ensuring the variables are used consistently in expressions and equations.
+
+It takes place after any evaluation of evaluable parameters and no evaluable parameters shall be evaluated during this process.
+(This must be true as long as tools allow the users to opt-out from unit-checking.)
+
+It is performed by using a type inference algorithm, albeit adapted to work with units.
+The algorithm proposed here is based on the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on the units of expressions.
+It extracts a sequence of substitutions from these constraints which once applied constructs the unit of a given expression.
+Any inconsistencies discovered during this process is a unit error, and any unit expression that can't be evaluated is considered undefined.
```

Comment:
1. Singular/plural
2. Trying to be more precise than _can't be evaluated_
```suggestion
Any inconsistency discovered during this process is a unit error, and any unit expression that contains remaining unknowns at the end of the process is considered undefined.
```

Reply:
https://www.masterclass.com/articles/any-singular-or-plural
The plural is fine here.